### DISCUSSION TITLE
the results from message passing and matrix multiplication are different

### DISCUSSION THREAD

**Discussion by zhf-0** [pyg-team/pytorch_geometric#4395] [Q&A] [answered]
I will illustrate my question by the following examples. Supposing the adjacent matrix is 
```
[  a_00     0       0
     0     a_11   a_12
     0     a_21   a_22 ]
```
so, the `edge_index` and `edge_weight` are
```
edge_index   = [    0        1       1        2         2
                    0        1       2        1         2   ]
edge_weigth = [   a_00      a_11    a_12    a_21      a_22  ]
```
If I use `GCNConv` and the node features are `x_0  x_1  x_2`, the tensor `x_j`  is (`x_j`  is source)
```
x_j = [            x_0      x_1     x_1     x_2        x_2   ]
```
According the code [GCNConv](https://pytorch-geometric.readthedocs.io/en/latest/_modules/torch_geometric/nn/conv/gcn_conv.html#GCNConv)
```
def message(self, x_j: Tensor, edge_weight: OptTensor) -> Tensor:
    return x_j if edge_weight is None else edge_weight.view(-1, 1) * x_j
```
The result is 
```
result = [ a_00 * x_0   a_11 * x_1  a_12 * x_1  a_21 *  x_2  a_22 *  x_2 ]
```
and updated features are 
```
0  =>  a_00 * x_0
1  =>  a_11 * x_1 + a_21 *  x_2
2  =>  a_12 * x_1 + a_22 *  x_2 
```
But, if I caculate the updated features through matrix multiplication
```
[  a_00     0       0        [ x_0           [ a_00 * x_0
     0     a_11   a_12         x_1     =>      a_11 * x_1 + a_12 * x_2    
     0     a_21   a_22 ]       x_2 ]           a_21 * x_1 + a_22 * x_2 ]
```
Only if the adjacent matrix is transposed, the updated features are the same. Why is this?

**Comment by rusty1s** [comment, marked as answer]
Yes, this is intended. The default message passing flow is from source nodes to target nodes (i.e. from `edge_index[0]` to `edge_index[1]`). You can change it by setting `flow="target_to_source"`. For dense GCN, this corresponds to `adj.t() @ x`. Since GCN is only properly defined for undirected graphs, this usually does not matter though.

**Comment by zhf-0** [comment]
Thank you very much!
